### PR TITLE
Add support for contiguous memory allocation and pointers

### PR DESCRIPTION
**Pull request by GlowingScrewdriver**
The goal is to have support for multi-dimensional arrays and pointers.
* memory is allocated on the stack
* `alloc` and `load` are implemented as value instructions
* `store` instruction takes two arguments and returns nothing
* _More notes will be added as progress is made_

The following tests are ported and passing from BRIL:
* `ptr_call` -- pointer as a function parameter
* `ptr_ret` -- pointer as a return value
* `access`
* `access_many`
* `access_ptr` -- multi-dimensional pointers
* `alloc`
* `fib` -- computing the Fibonacci sequence

The following tests will not be ported:
* `alloc_large`, `alloc_many` -- repeated allocation and freeing. Not applicable since we don't have the concept of freeing.

### CODE REVIEW

**Comment by GlowingScrewdriver** [comment]
Ik I forgot to apply Black :grimacing:
Will do for the next commit

**Comment by GlowingScrewdriver** [comment]
Added test: `mem_id`
Also implemented the `id` instruction (which is actually a core BRIL instruction)

**Comment by chsasank** [comment]
Do run black

File: src/backend/brilisp.scm [hunk 1]
```diff
@@ -34,12 +34,17 @@
...
            (instrs (rest (rest expr))))
         `((name . ,name)
-          (type . ,type)
+          (type . ,(gen-type type))
           (args . ,(map-vec gen-arg args))
           (instrs . ,(map-vec gen-instr instrs)))))
 
 (define (gen-arg arg)
-    `((name . ,(first arg)) (type . ,(second arg))))
+    `((name . ,(first arg)) (type . ,(gen-type (second arg)))))
```
Comment (chsasank) [inline src/backend/brilisp.scm:42]: Good to see recursion here

**Comment by chsasank** [comment]
This is a great PR. Clean code, lot of tests! Great job!

**Pull request merged**